deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,528 @@
+		(struct happy_connect_woker_args *)malloc(
+			sizeof(struct happy_connect_woker_args));
+	if (args == NULL) {
+		context->error = ENOMEM;
+		context->error_message =
+			"happy-eyeballs: Failed to allocate memory for worker context";
+		return STATUS_FAILURE;
+	}
+	args->sockfd = (int)fd;
+	args->address = addr;
+	args->context = context;
+	args->candidate = candidate;
+
+	int result = os_event_init(&candidate->socket_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		// failure to create the
+		context->error = result;
+		context->error_message =
+			"happy-eyeballs: Failed to initialize socket completed event";
+		return STATUS_FAILURE;
+	}
+
+	result = pthread_create(&candidate->thread, NULL, happy_connect_worker,
+				args);
+	if (result != 0) {
+		// failure to start the worker thread
+		context->error = result;
+		context->error_message = strerror(result);
+		os_event_destroy(candidate->socket_completed_event);
+		return STATUS_FAILURE;
+	}
+
+	candidate->valid = true;
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_init(struct happy_eyeballs_ctx **context)
+{
+	if (context == NULL) {
+		return STATUS_INVALID_ARGUMENT;
+	}
+
+	*context = (struct happy_eyeballs_ctx *)malloc(
+		sizeof(struct happy_eyeballs_ctx));
+	if (*context == NULL) {
+		return STATUS_FAILURE;
+	}
+
+	memset(*context, 0, sizeof(struct happy_eyeballs_ctx));
+
+	// race_completed_event will be signalled when there is a winner
+	int result = os_event_init(&(*context)->race_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		(*context)->error = result;
+		(*context)->error_message =
+			"happy-eyeballs: Failed to initialize race completed event";
+		return STATUS_FAILURE;
+	}
+
+	// this mutex is used to avoid the situation where we may have two simultaneous winners and inconsistent
+	// values set to the context.
+	result = pthread_mutex_init(&(*context)->winner_mutex, NULL);
+	if (result != 0) {
+		(*context)->error = result;
+		(*context)->error_message =
+			"happy-eyeballs: Failed to initialize winner mutex";
+		return STATUS_FAILURE;
+	}
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_connect(struct happy_eyeballs_ctx *context,
+			   const char *hostname, int port)
+{
+	if (hostname == NULL || context == NULL || port == 0) {
+		return STATUS_INVALID_ARGUMENT;
+	}
+
+	int result = check_comodo(context);
+	if (result != STATUS_SUCCESS) {
+		return result;
+	}
+
+	result = build_addr_list(hostname, port, context);
+	if (result != STATUS_SUCCESS) {
+		return result;
+	}
+
+	int prev_family = 0;
+	struct addrinfo *next = context->addresses;
+
+	// Exit the loop under the following conditions:
+	//	1. We have reached the maximum allowed number of attempts
+	//	2. There are no more candidates in the linked list (ie. `next` is null)
+	//  3. We have seen two candidates of the same family in a row, stop happy eyeballs and let the previous attempt go it alone.
+	for (int i = 0; i < HAPPY_EYEBALLS_MAX_ATTEMPTS && next &&
```

Comment:
I think you mentioned restructuring to use darray, but can simplify iteration over tehse